### PR TITLE
kernel/will-it-scale: Add scalability test

### PR DESCRIPTION
**Pull request by sacsant**
This patch adds a test to validate scalability aspect of kernel. Will It Scale
takes a testcase and runs n parallel copies to see if the testcase will scale. 
This test is based on https://github.com/antonblanchard/will-it-scale

The test includes yaml based input for each supported testcase.

Signed-off-by: Sachin Sant <sachinp@linux.ibm.com>

### CODE REVIEW

**Comment by PraveenPenguin** [comment]
@sacsant can see scope to merge all yaml in single YAML as it will reduce maintenance effort

**Comment by abdhaleegit** [comment]
> @sacsant can see scope to merge all yaml in single YAML as it will reduce maintenance effort

single yaml with all test mux variants

**Comment by sacsant** [comment]
Depending on resources available on system, each test takes anywhere between 40min to a couple of hours. I don't think we should run the test against all option available as it will take maybe more than 24 hours.

I can include a couple of sample YAML files and remove the remaining one. The README file has information about all possible tests. A user who wants to run a specific test can always create their own yaml's.  

Would that work?

**Comment by sacsant** [comment]
I have consolidated the YAML's into a single one using mux and have added a warning about long execution time.

**Comment by sacsant** [comment]
Test result 
 (1/1) kernel/will-it-scale.py:WillItScaleTest.test_scaleitall;run-testcase-mmap1-f366: STARTED
 (1/1) kernel/will-it-scale.py:WillItScaleTest.test_scaleitall;run-testcase-mmap1-f366: PASS (1590.29 s)
RESULTS    : PASS 1 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0

File: kernel/will-it-scale.py.data/will-it-scale.yaml [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+#testcase: !mux
+#    All:
+#        name: All
+#        postprocess: True
+# Each testcase can take upto an hour or so depending on amount
+# of resources available on the system under test
+testcase: !mux
+    brk1:
+        name: brk1
+        postprocess: True
```
Comment (abdhaleegit) [inline kernel/will-it-scale.py.data/will-it-scale.yaml:11]: remove  postprocess: True  in each variant and keep it above testcase: !mux  makes it available for each variant

**Review by abdhaleegit — CHANGES_REQUESTED** [review]

File: kernel/will-it-scale.py [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+                                   expire='7d')
+        archive.extract(tarball, self.workdir)
+        self.sourcedir = os.path.join(self.workdir, 'will-it-scale-master')
+        os.chdir(self.sourcedir)
+        if build.make(self.sourcedir):
+            self.fail('make failed, please check debug logs')
+
+    @skipIf(SINGLE_NODE, "Test requires atleast two numa nodes")
+    @skipIf(VERSION_CHK, "Test requires Python 3.7+")
+    def test_scaleitall(self):
```
Comment (PraveenPenguin) [inline kernel/will-it-scale.py:71]: as having single test move skip statement in setup itself else unnecessary setUp will call

File: kernel/will-it-scale.py [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+        """
+        To execute test using git copy
+          make
+          ./runalltests
+
+        To generate graphical results
+          ./postprocess.py
+        """
+        smm = SoftwareManager()
+        pkgs = ['gcc', 'make', 'hwloc-devel']
```
Comment (PraveenPenguin) [inline kernel/will-it-scale.py:51]: is `hwloc-devel` package available on all distro , as i am locate this only in RHEL9.*/centos9 onward ? Please double-check and consider test case in other distro if test not supported or we can have other alternative of this package ,

File: kernel/will-it-scale.py [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+        self.sourcedir = os.path.join(self.workdir, 'will-it-scale-master')
+        os.chdir(self.sourcedir)
+        if build.make(self.sourcedir):
+            self.fail('make failed, please check debug logs')
+
+    @skipIf(SINGLE_NODE, "Test requires atleast two numa nodes")
+    @skipIf(VERSION_CHK, "Test requires Python 3.7+")
+    def test_scaleitall(self):
+        """
+        Invoke and execute all tests
```
Comment (PraveenPenguin) [inline kernel/will-it-scale.py:73]: Can it will reword as `execute test(s)` as it running test based on input

File: kernel/will-it-scale.py [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+
+        # Identify the test to be executed
+        if self.testcase in 'All':
+            cmd = './runalltests'
+        else:
+            cmd = './runtest.py %s > %s.csv' % (self.testcase, self.testcase)
+
+        # Execute the test(s)
+        ret = process.system(cmd, shell=True, sudo=True, ignore_status=True)
+        if ret != 0:
```
Comment (PraveenPenguin) [inline kernel/will-it-scale.py:86]: Please consider like 

`if process.system(cmd, shell=True, sudo=True, ignore_status=True) !=0 :`

as it avoid to have new variable

File: kernel/will-it-scale.py [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+            self.fail('Please check the logs for failure')
+        if self.testcase not in 'All':
+            self.filetocopy = self.testcase + ".csv"
```
Comment (PraveenPenguin) [inline kernel/will-it-scale.py:89]: Can we use string concatenation ?

File: kernel/will-it-scale.py [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+            shutil.copy(self.filetocopy, self.logdir)
+
+        # Generate graphical results if postprocessing is enabled
+        if self.postprocess:
+            cmd = './postprocess.py'
+            ret = process.system(cmd, shell=True, ignore_status=True)
+            if ret != 0:
```
Comment (PraveenPenguin) [inline kernel/will-it-scale.py:96]: here also please 

`if process.system(cmd, shell=True, ignore_status=True) != 0:`

File: kernel/will-it-scale.py.data/README [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+ - tlb_flush3
+ - unix1
+ - unlink1
+ - unlink2
+ - write1
+ - writeseek1
+ - writeseek2
+ - writeseek3
+
+Please refer to the github project documentation on how to add new test cases.
```
Comment (PraveenPenguin) [inline kernel/will-it-scale.py.data/README:70]: is this statement relevant to the current repository ? It should  not like 


Please refer to the GitHub project documentation to pick other test cases

File: kernel/will-it-scale.py.data/will-it-scale.yaml [hunk 9]
```diff
@@ -0,0 +1,129 @@
+# To run all tests use following stanza instead
+#testcase: !mux
+#    All:
+#        name: All
+#        postprocess: True
```
Comment (PraveenPenguin) [inline kernel/will-it-scale.py.data/will-it-scale.yaml:5]: I feel these comment statement can be removed as user already knowing how to use ? May this can move as one of combination

File: kernel/will-it-scale.py [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+        To execute test using git copy
+          make
+          ./runalltests
+
+        To generate graphical results
+          ./postprocess.py
+        """
+        smm = SoftwareManager()
+        pkgs = ['gcc', 'make', 'hwloc-devel']
+        for package in pkgs:
```
Comment (PraveenPenguin) [inline kernel/will-it-scale.py:52]: just have like 

`for package in ['gcc', 'make', 'hwloc-devel']: `

**Review by PraveenPenguin — CHANGES_REQUESTED** [review]

File: kernel/will-it-scale.py (hunk 2 above)
Comment (sacsant) [reply kernel/will-it-scale.py:70]: Have moved the skipif statement to setup()

File: kernel/will-it-scale.py.data/will-it-scale.yaml (hunk 9 above)
Comment (sacsant) [reply kernel/will-it-scale.py.data/will-it-scale.yaml]: Yes, have removed it from yaml and added it to README.

**Comment by sacsant** [comment]
Test run after review comments have been incorporated.

 (1/1) kernel/will-it-scale.py:WillItScaleTest.test_scaleitall: STARTED
 (1/1) kernel/will-it-scale.py:WillItScaleTest.test_scaleitall: PASS (795.09 s)

**Review by abdhaleegit — APPROVED** [review]
LGTM

File: kernel/will-it-scale.py (hunk 3 above)
Comment (PraveenPenguin) [reply kernel/will-it-scale.py]: @sacsant thanks for addressing review comments, can you please look in this comment as well

Comment (sacsant) [reply kernel/will-it-scale.py]: Ah, thanks for the reminder. I missed this one. I checked RHEL & Ubuntu and the package is available. SLES does not seem to have this package. I can add a check for distro versions but was wondering since the test gets cancelled anyway if the package is not available, should we add this extra code?

**Review by PraveenPenguin — APPROVED** [review]
